### PR TITLE
Unique entity custom message

### PR DESCRIPTION
**Pull request by jamyouss**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

Add the possibility to have custom message for unique entity constraint.
For example if we have a collection of entities in a form, the message can be more expressive.

Example:

```
/**
 * Town
 *
 * @ORM\Entity()
 * @UniqueEntity(fields={"city"}, message="The city '{{ city }}' is already used.")
 */
class Town
{
```

I accidentally closed the previous opened PR (#13787), that why a recreate a pull request.

### CODE REVIEW

**Comment by Minishlink** [comment]
Thanks :+1:

**Comment by iltar** [comment]
Is the twig style standard or the translations style standard? `{{ var }}` vs `%var%`.

**Comment by Minishlink** [comment]
According to other uses of variables in constraints messages (eg. EmailValidator), the standard is twig-style.

**Comment by oleg-andreyev** [comment]
All violations messages are passed to `translator` where placeholders are replaced (see `\Symfony\Component\Translation\Translator::trans`), so your value must be added to `$parameters` and let `translator` do it's job.

For example see: `EmailValidator`

**Comment by keradus** [comment]
no new tests needed for that new behavior ?

**Comment by OskarStark** [comment]
great work! :+1:

**Comment by razmo** [comment]
@keradus I don't think so because i don't change the validation behaviour.

**Comment by keradus** [comment]
OK, but how to we know that new behaviour works after some time? Tests can prove that when sb refactor sth this behaviour will still works ;)

**Comment by SofHad** [comment]
Good :+1:

**Comment by razmo** [comment]
@keradus you're right, i gonna add some tests

**Comment by OskarStark** [comment]
thank you @razmo ,give me a ping an i'll review it

**Comment by razmo** [comment]
Your welcome @OskarStark ! 
What does "give me a ping" mean ? (I'm not used to contribute to projects on github)

**Comment by OskarStark** [comment]
if you use ```@OskarStark``` i get a direct notification via mail, as you did ;-)

**Comment by razmo** [comment]
@OskarStark test updated

**Comment by OskarStark** [comment]
hmm, I don't know why the travis build is broken with 5.6, can you have a look @razmo

**Comment by razmo** [comment]
@OskarStark I run phpunit test in local and i have 5 failure on ```Symfony\Component\HttpKernel\Tests\HttpCache\HttpCacheTest``` and 1 failure on ```Symfony\Component\Finder\Tests\Iterator\RecursiveDirectoryIteratorTest::testSeek```

I don't think that it is caused by my changes

**Comment by OskarStark** [comment]
hmm, ok

ping @xabbuh @stof

File: src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php [hunk 1]
```diff
@@ -134,6 +135,16 @@ public function validate($entity, Constraint $constraint)
         $errorPath = null !== $constraint->errorPath ? $constraint->errorPath : $fields[0];
         $invalidValue = isset($criteria[$errorPath]) ? $criteria[$errorPath] : $criteria[$fields[0]];
 
+        $vars = array();
+
+        if (preg_match_all('/{{ ([a-zA-Z0-9_]+) }}/i', $constraint->message, $vars) > 0) {
+            $accessor = PropertyAccess::createPropertyAccessor();
```
Comment (xabbuh) [inline src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php]: You need to make sure that the PropertyAccess component is actually available and you should add it to the `suggest` section in `composer.json`.

File: src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php [hunk 2]
```diff
@@ -134,6 +135,16 @@ public function validate($entity, Constraint $constraint)
         $errorPath = null !== $constraint->errorPath ? $constraint->errorPath : $fields[0];
         $invalidValue = isset($criteria[$errorPath]) ? $criteria[$errorPath] : $criteria[$fields[0]];
 
+        $vars = array();
+
+        if (preg_match_all('/{{ ([a-zA-Z0-9_]+) }}/i', $constraint->message, $vars) > 0) {
```
Comment (xabbuh) [inline src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php]: The `i` modifier does not seem to be necessary when you list lowercased and uppercased letters in the character class.

File: src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php [hunk 3]
```diff
@@ -134,6 +135,16 @@ public function validate($entity, Constraint $constraint)
...
+            $accessor = PropertyAccess::createPropertyAccessor();
+
+            foreach ($vars[1] as $var) {
+                $constraint->message = str_replace(sprintf('{{ %s }}', $var), $accessor->getValue($entity, $var), $constraint->message);
```
Comment (stof) [inline src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php]: please don't make replacements yourselves. It makes the message incompatible with usage of the translator. Replacements must be applied after the translator is used

**Comment by xabbuh** [comment]
Wouldn't it be better to be able to configure a list of object property names whose value would be passed to the actual constraint through the `setParameter()` method of the constraint violation builder?

**Comment by razmo** [comment]
Thank you @stof and @xabbuh for those feedbacks. I going to update as soon as possible.

**Comment by razmo** [comment]
@stof @xabbuh : PR updated !
My PR is based on the 2.6 branch. Do i need to merge the 2.7 into my branch ?

**Comment by maryo** [comment]
What about also passing the valid entity/entities returned by the repository to the constraint violation?
https://github.com/symfony/doctrine-bridge/blob/master/Validator/Constraints/UniqueEntityValidator.php#L115

Usecase: I want to show a hint for user in a registration form when she wants to register with an email of an already registered but not activated account. In this case i need to have an access to the user in the DB. OFC I can write it myself in a controller or using a form event. But this way it could be just in the template which i prefer when overriding templates of thirdparty bundle. I have similar logic in login template but there is an access to the securityexception which contains the user.

Something like "This email is already used but it's account is not activated yet. Click [here] to activate the account. "

**Comment by razmo** [comment]
ping @stof

**Comment by OskarStark** [comment]
@razmo the merge to the 2.7 branch is done automatically afterwards

**Comment by razmo** [comment]
@OskarStark Ok thx

**Comment by razmo** [comment]
@OskarStark Phpunit tests passed in my local but didn't pass on travis. Have you any idea why ?

**Comment by OskarStark** [comment]
no sorry @razmo

**Comment by xabbuh** [comment]
Failures seem to be related to the fact that your changes are based off the `2.6` branch.

**Comment by razmo** [comment]
Wired because because my branch ```uniqueEntityCustomMessage``` is sync with the ```2.6``` branch

**Comment by xabbuh** [comment]
@razmo Sure, but Symfony 2.6 itself doesn't receive bug fixes anymore. Probably, you can better reopen the PR based on the `2.8` branch.

**Comment by razmo** [comment]
Close this PR and reopen it based on 2.8 branch

**Pull request closed without merging**